enhancements/olm-deletes-operands.md

```diff
@@ -0,0 +1,155 @@
+Once the CRs have their finalizers cleared by the operator and are removed from the foreground, OLM will delete the CRD and then proceed with the removal of the operator.
+
+```YAML
+apiVersion: operators.coreos.com/v1alpha1
+kind: Subscription
+spec:
+  ...
+  operandCleanup:
+    type: olm-managed
+    timeout: 30s
+```
+
+Additionaly a timeout configured by `spec.operandCleanup.timeout` would specify how long OLM should wait for the operator to finish cleaning up.
+
+If the cleanup timeout is exceeded, OLM should generate an alert event indicating a cleanup failure for that operator, and then proceed to uninstall the operator.
+
+**Note:** See [Risks and Mitigations section][risks-section] on configuring the timeout.
+
+#### Cleanup for namespace-scoped operators
+
+If there are two non-global operators installed of the same type (providing the same APIs), then removing the CRD during the cleanup of one operator would result in the deletion of CRs for the other operator as well.
+
+To avoid this case OLM would have to ensure that it only removes the CRD during cleanup if:
+
+- The operator installed is global
+- The non-global operator is the only installation or `Subscription` for that operator type in the cluster
+
+#### Operator managed cleanup (not in scope for this proposal)
+
+There could be a use case where operator authors might require custom logic for deleting CRs e.g to order the deletion, or filter certain CRs from deletion to preserve state.
+
+For that `spec.operandCleanup` could be extended to another type, e.g `type: operator-managed`, which could allow a user to configure a job with custom logic for the deletion of CRs.
+The exact workflow and UX are unclear at this time, and are left as a future extension.
+
+## Design Details
+
+### Test Plan
+
+OLM's e2e testing suite should be upgraded to handle the following use cases:
+
+- An operator that is uninstalled with `spec.operandCleanup` unspecified or `none` preserves the CRs and CRD
+- An operator that is uninstalled with `spec.operandCleanup: olm-managed` removes all CRs and CRDs
+- An operator that is uninstalled with `spec.operandCleanup: olm-managed` and exceeds a timeout should generate an alert event, and not block operator removal
+
+### Risks and Mitigations
+
+It's worth considering if the clean up timeout `spec.operandCleanup.timeout` should have a "sane" upper bound to avoid having the operator uninstall be blocked by cleanup indefinitely.
+Or if users would require setting an inifinte timeout to block operator removal, if they want to inspect operator logs in the event of a cleanup failure.
+
+Other process, e.g cluster deletion, waiting for operators to finish clean up could end up being blocked indefinitely or for a long time in this case.
+
+## Drawbacks
+
+Goes against OLM's conservative approach to not remove APIs in order to prevent data loss. However this feature is opt-in so it shouldn't be a risk in the default setting.
```

Comment:
It's opt-in, but it's also a frequent point of support / feature request. We may want a corresponding feature in the console that suggests this be set before an operator is removed (would be neat to think of a way to do something similar for kubectl, but nothing springs to mind).

Reply:
Our `kubectl plugin` comes to mind.

Reply:
Also a UI could always opt-in and highlight this behavior visually in a deletion modal with a checkbox ticked by default.

Reply:
This absolutely needs to be part of the UI.  It may not be there the day OLM ships this feature, but please work w/ the console team to add a ticket to their board to enable this as a first class UI option (both "opt into cleanup" and "override/force cleanup even though the operator doesn't support it")

@spadgett fyi.